contracts/reputation/Reputation.sol

```diff
@@ -14,81 +14,289 @@ pragma solidity >=0.5.0 <0.6.0;
+     * @notice Increases reputation of a validator by delta.
+     *
+     * @dev Function requires
+     *          - only consensus can call
+     *          - the specified validator is active
+     *
+     * @param _validator A validator for which to increase a reputation.
+     * @param _delta A change (delta) to increase a validator's reputation.
+     *
+     * @return Returns an updated reputation.
+     */
+    function increase(address _validator, uint256 _delta)
+        external
+        onlyConsensus
+        isActive(_validator)
+        returns (uint256)
+    {
+        reputations[_validator] = reputations[_validator].add(_delta);
+
+        return reputations[_validator];
     }
 
-    function join(address _validator)
+    /**
+     * @notice Decreases reputation of a validator by delta.
+     *         Sets the validator's reputation to 0 if after operation
+     *         it's negative.
+     *
+     * @dev Function requires:
+     *          - only consensus can call.
+     *          - the specified validator is active.
+     *
+     * @param _validator A validator for which to decrease a reputation.
+     * @param _delta A change (delta) to decrease a validator's reputation.
+     *
+     * @return Returns an updated reputation.
+     */
+    function decrease(address _validator, uint256 _delta)
+        external
         onlyConsensus
+        isActive(_validator)
+        returns (uint256)
+    {
+        if (reputations[_validator] < _delta) {
+            reputations[_validator] = 0;
+        } else {
+            reputations[_validator] = reputations[_validator].sub(_delta);
+        }
+
+        return reputations[_validator];
+    }
+
+    /**
+     * @dev Function requires:
+     *          - only consensus can call
+     *          - a validator was not joined previously
+     *          - a validator address is not 0
+     *          - a withdrawal address is not 0
+     *          - a validator address is not same as its withdrawal address
+     *          - a validator approved in mOST token contract to transfer
+     *            a stake amount.
+     *          - a validator approved in wETH token contract to transfer
+     *            a stake amount.
+     */
+    function join(
+        address _validator,
+        address _withdrawalAddress
+    )
         external
-        view
+        onlyConsensus
     {
-        require(status[_validator] == ValidatorStatus.Undefined,
-            "No validator can rejoin.");
-        // continue
+        require(
+            _validator != address(0),
+            "Validator address is 0."
+        );
+
+        require(
+            _withdrawalAddress != address(0),
+            "Validator's withdrawal address is 0."
+        );
+
+        require(
+            _validator != _withdrawalAddress,
+            "Validator's address is the same as its withdrawal address."
+        );
+
+        require(
+            statuses[_validator] == ValidatorStatus.Undefined,
```

Comment:
perhaps worth checking that the withdrawal address is also not in the statuses list; a weird edge-case, but better to check

Reply:
also check withdrawal addresses is not in `withdrawaladdresses`

Reply:
Done.